ml/randomforest.h

```diff
@@ -0,0 +1,657 @@
+#include "mldb/base/parallel.h"
+#include "mldb/base/thread_pool.h"
+#include "mldb/server/column_scope.h"
+#include "mldb/server/bucket.h"
+
+namespace Datacratic {
+namespace MLDB {
+
+/** Holds the set of data for a partition of a decision tree. */
+struct PartitionData {
+
+    PartitionData()
+        : fs(nullptr)
+    {
+    }
+
+    PartitionData(const DatasetFeatureSpace & fs)
+        : fs(&fs), features(fs.columnInfo.size())
+    {
+        for (auto & c: fs.columnInfo) {
+            Feature & f = features.at(c.second.index);
+            f.active = c.second.distinctValues > 1;
+            f.buckets = c.second.buckets;
+            f.info = &c.second;
+            f.ordinal = c.second.bucketDescriptions.isOnlyNumeric();
+        }
+    }
+
+    /** Create a new dataset with the same labels, different weights
+        (by element-wise multiplication), and with
+        zero weights filtered out such that example numbers are strictly
+        increasing.
+    */
+    PartitionData reweightAndCompact(const std::vector<float> & weights) const
+    {
+        size_t numNonZero = 0;
+        for (auto & w: weights)
+            numNonZero += (w != 0);
+        
+        PartitionData data;
+        data.features = this->features;
+        data.fs = this->fs;
+        data.reserve(numNonZero);
+
+        std::vector<WritableBucketList>
+            featureBuckets(features.size());
+
+        for (unsigned i = 0;  i < data.features.size();  ++i) {
+            if (data.features[i].active) {
+                featureBuckets[i].init(numNonZero,
+                                       data.features[i].info->distinctValues);
+
+                data.features[i].buckets = featureBuckets[i];
+            }
+        }
+
+        auto doFeature = [&] (size_t f)
+            {
+                if (f == data.features.size()) {
+                    // Do the row index
+                    size_t n = 0;
+                    for (size_t i = 0;  i < rows.size();  ++i) {
+                        if (weights[i] == 0)
+                            continue;
+                        data.addRow(rows[i].label, rows[i].weight * weights[i],
+                                    n++);
+                    }
+                    ExcAssertEqual(n, numNonZero);
+                    return;
+                }
+
+                if (!data.features[f].active)
+                    return;
+
+                size_t n = 0;
+                for (size_t i = 0;  i < rows.size();  ++i) {
+                    if (weights[i] == 0)
+                        continue;
+
+                    uint32_t bucket = features[f].buckets[rows[i].exampleNum];
+                    featureBuckets[f].write(bucket);
+                    ++n;
+                }
+
+                ExcAssertEqual(n, numNonZero);
+            };
+
+        Datacratic::parallelMap(0, data.features.size() + 1, doFeature);
+
+        return data;
+    }
+
+    const DatasetFeatureSpace * fs;
```

Comment:
Here we keep a pointer to a const ref passed at construction.  Would it be safer and clearer to use a shared pointer?  Also, if it stays this way we should mentioned that PartitionData stores a pointer to the feature space and therefore the scope of the feature space must be as long as the one of PartitionData.